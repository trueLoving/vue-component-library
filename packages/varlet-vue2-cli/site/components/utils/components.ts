import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>
  unmount(): void
}

export function mountComponent(component: ComponentOptions<Vue>, container = 'body', options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}
