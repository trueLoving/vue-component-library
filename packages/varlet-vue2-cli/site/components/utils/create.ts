import Vue, { ComponentOptions, RenderContext, VueConstructor } from 'vue'
import { SlotsMixin } from './mixins/slots'
import { ListenersMixin } from './mixins/listeners'

export type ComponentOptionsPlugin = ComponentOptions<Vue> & { install?: (app: VueConstructor) => void }

export function nextTick() {
  return new Promise(Vue.nextTick)
}

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function defineComponent(sfc: ComponentOptionsPlugin & { name: string }) {
  if (sfc.mixins) {
    sfc.mixins.push(SlotsMixin)
    sfc.mixins.push(ListenersMixin)
  } else {
    sfc.mixins = [SlotsMixin, ListenersMixin]
  }

  return sfc
}
