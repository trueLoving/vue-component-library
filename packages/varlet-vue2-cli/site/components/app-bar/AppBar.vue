<template>
  <div
    class="var-site-app-bar"
    :class="{ 'var-site-elevation--3': elevation }"
    :style="{
      background: color,
      color: textColor,
    }"
  >
    <div class="var-site-app-bar__left">
      <slot name="left" />
      <div class="var-site-app-bar__title" :style="{ paddingLeft }" v-if="titlePosition === 'left'">
        <slot>{{ title }}</slot>
      </div>
    </div>

    <div class="var-site-app-bar__title" v-if="titlePosition === 'center'">
      <slot>{{ title }}</slot>
    </div>

    <div class="var-site-app-bar__right">
      <div class="var-site-app-bar__title" :style="{ paddingRight }" v-if="titlePosition === 'right'">
        <slot>{{ title }}</slot>
      </div>
      <slot name="right" />
    </div>
  </div>
</template>

<script>
import { defineComponent } from '../utils/create'
import { props } from './props'

export default defineComponent({
  name: 'VarSiteAppBar',

  props,

  data: () => ({
    paddingLeft: 0,
    paddingRight: 0,
  }),

  methods: {
    computePadding() {
      this.paddingLeft = this.hasSlots('left') ? 0 : undefined
      this.paddingRight = this.hasSlots('right') ? 0 : undefined
    },
  },

  mounted() {
    this.computePadding()
  },

  updated() {
    this.computePadding()
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import './appBar';
</style>
