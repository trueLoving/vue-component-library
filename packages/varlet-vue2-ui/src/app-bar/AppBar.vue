<template>
  <div
    class="var-app-bar"
    :class="{ 'var-elevation--3': elevation }"
    :style="{
      background: color,
      color: textColor,
    }"
  >
    <div class="var-app-bar__left">
      <slot name="left" />
      <div class="var-app-bar__title" :style="{ paddingLeft }" v-if="titlePosition === 'left'">
        <slot>{{ title }}</slot>
      </div>
    </div>

    <div class="var-app-bar__title" v-if="titlePosition === 'center'">
      <slot>{{ title }}</slot>
    </div>

    <div class="var-app-bar__right">
      <div class="var-app-bar__title" :style="{ paddingRight }" v-if="titlePosition === 'right'">
        <slot>{{ title }}</slot>
      </div>
      <slot name="right" />
    </div>
  </div>
</template>

<script>
import { defineComponent } from '../utils/create'
import { props } from './props'

export default defineComponent({
  name: 'VarAppBar',

  props,

  data: () => ({
    paddingLeft: 0,
    paddingRight: 0,
  }),

  mounted() {
    this.computePadding()
  },

  updated() {
    this.computePadding()
  },

  methods: {
    computePadding() {
      this.paddingLeft = this.hasSlots('left') ? 0 : undefined
      this.paddingRight = this.hasSlots('right') ? 0 : undefined
    },
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import './appBar';
</style>
