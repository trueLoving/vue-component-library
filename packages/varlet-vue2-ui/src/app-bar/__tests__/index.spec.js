import example from '../example'
import AppBar from '..'
import VarAppBar from '../AppBar'
import { mount } from '@vue/test-utils'
import Vue from 'vue'
import { delay } from '../../utils/jest'

test('test app bar example', async () => {
  const wrapper = mount(example)
  await delay(100)

  expect(wrapper.html()).toMatchSnapshot()
})

test('test app bar plugin', () => {
  Vue.use(AppBar)
  expect(Vue.component(AppBar.name)).toBeTruthy()
})

test('test app bar props', () => {
  const wrapper = mount(VarAppBar, {
    propsData: {
      color: '#a3b9db',
      textColor: '#000',
      title: 'test title',
      titlePosition: 'center',
      elevation: false,
    },
    scopedSlots: {
      left: () => 'leftSlot',
      right: () => 'rightSlot',
    },
  })

  expect(wrapper.html()).toMatchSnapshot()
})
