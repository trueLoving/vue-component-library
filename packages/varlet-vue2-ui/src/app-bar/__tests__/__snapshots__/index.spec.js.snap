// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test app bar example 1`] = `
"<div class=\\"var-appBar-example\\">
  <div class=\\"app-type\\">基础导航栏</div>
  <div class=\\"var-app-bar var-elevation--3\\">
    <div class=\\"var-app-bar__left\\">
      <div class=\\"var-app-bar__title\\" style=\\"padding-left: 0px;\\">标题</div>
    </div>
    <!---->
    <div class=\\"var-app-bar__right\\">
      <!---->
    </div>
  </div>
  <div class=\\"app-type\\">自定义样式</div>
  <div class=\\"var-app-bar var-elevation--3\\" style=\\"background: rgb(0, 196, 143);\\">
    <div class=\\"var-app-bar__left\\">
      <!---->
    </div>
    <div class=\\"var-app-bar__title\\">标题</div>
    <div class=\\"var-app-bar__right\\">
      <!---->
    </div>
  </div>
  <div class=\\"app-type\\">添加标题处插槽</div>
  <div class=\\"var-app-bar var-elevation--3\\">
    <div class=\\"var-app-bar__left\\">
      <div class=\\"var-app-bar__title\\" style=\\"padding-left: 0px;\\">从插槽处添加标题</div>
    </div>
    <!---->
    <div class=\\"var-app-bar__right\\">
      <!---->
    </div>
  </div>
  <div class=\\"app-type\\">添加左侧插槽</div>
  <div class=\\"var-app-bar var-elevation--3\\">
    <div class=\\"var-app-bar__left\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--round\\" style=\\"color: rgb(255, 255, 255); background: transparent;\\">
        <!---->
        <div class=\\"var-button__content\\"><i class=\\"var-icon var-icon--set var-icon-chevron-left\\" style=\\"transition: transform 0ms; font-size: 24px;\\"></i></div>
      </button>
      <!---->
    </div>
    <div class=\\"var-app-bar__title\\">标题</div>
    <div class=\\"var-app-bar__right\\">
      <!---->
    </div>
  </div>
  <div class=\\"app-type\\">添加右侧插槽</div>
  <div class=\\"var-app-bar var-elevation--3\\">
    <div class=\\"var-app-bar__left\\">
      <div class=\\"var-app-bar__title\\" style=\\"padding-left: 0px;\\">标题</div>
    </div>
    <!---->
    <div class=\\"var-app-bar__right\\">
      <!----> <button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--round\\" style=\\"color: rgb(255, 255, 255); background: transparent;\\">
        <!---->
        <div class=\\"var-button__content\\"><i class=\\"var-icon var-icon--set var-icon-magnify\\" style=\\"transition: transform 0ms; font-size: 24px;\\"></i></div>
      </button></div>
  </div>
  <div class=\\"app-type\\">添加左右插槽</div>
  <div class=\\"var-app-bar var-elevation--3\\">
    <div class=\\"var-app-bar__left\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--round\\" style=\\"color: rgb(255, 255, 255); background: transparent;\\">
        <!---->
        <div class=\\"var-button__content\\"><i class=\\"var-icon var-icon--set var-icon-chevron-left\\" style=\\"transition: transform 0ms; font-size: 24px;\\"></i></div>
      </button>
      <div class=\\"var-app-bar__title\\" style=\\"padding-left: 0px;\\">标题</div>
    </div>
    <!---->
    <div class=\\"var-app-bar__right\\">
      <!---->
      <div class=\\"var-menu\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--round\\" style=\\"color: rgb(255, 255, 255); background: transparent;\\">
          <!---->
          <div class=\\"var-button__content\\"><i class=\\"var-icon var-icon--set var-icon-menu\\" style=\\"transition: transform 0ms; font-size: 24px;\\"></i></div>
        </button>
        <div class=\\"var-teleport\\"></div>
      </div>
    </div>
  </div>
  <div class=\\"space\\"></div>
</div>"
`;

exports[`test app bar props 1`] = `
"<div class=\\"var-app-bar\\" style=\\"background: rgb(163, 185, 219); color: rgb(0, 0, 0);\\">
  <div class=\\"var-app-bar__left\\">leftSlot
    <!---->
  </div>
  <div class=\\"var-app-bar__title\\">test title</div>
  <div class=\\"var-app-bar__right\\">
    <!----> rightSlot</div>
</div>"
`;
